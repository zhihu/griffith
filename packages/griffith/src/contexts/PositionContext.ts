import React from 'react'

type PositionContextValue = {
  isFullWidth: boolean
  helperImageSrc?: string | null
  updateVideoSize: (opts: {videoWidth: number; videoHeight: number}) => void
}

const PositionContext = React.createContext<PositionContextValue>({
  isFullWidth: false,
  updateVideoSize() {},
})
PositionContext.displayName = 'PositionContext'

export default PositionContext
