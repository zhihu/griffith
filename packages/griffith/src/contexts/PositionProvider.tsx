import React from 'react'
import {css} from 'aphrodite/no-important'
import {reduce} from 'griffith-utils'
import listenResize from '../utils/listenResize'
import PositionContext from './PositionContext'
import styles from './PositionProvider.styles'

type Props = {
  shouldObserveResize?: boolean
}

type State = any

export default class PositionProvider extends React.PureComponent<
  Props,
  State
> {
  unlistenResize_: any

  state = {
    videoWidth: 0,
    videoHeight: 0,
    isFullWidth: false,
    helperImageSrc: null,
  }

  ref = React.createRef()

  componentDidMount() {
    if (this.props.shouldObserveResize) {
      this.startObservingResize()
    }
    this.updateHelperImageSrc()
  }

  componentDidUpdate(prevProps: Props, prevState: State) {
    const {shouldObserveResize: prevShouldObserve} = prevProps
    const {videoWidth: prevWidth, videoHeight: prevHeight} = prevState
    const {shouldObserveResize} = this.props
    const {videoWidth, videoHeight} = this.state

    if (prevWidth !== videoWidth || prevHeight !== videoHeight) {
      this.triggerUpdateIsFullWidth()
      this.updateHelperImageSrc()
    }

    if (!prevShouldObserve && shouldObserveResize) {
      this.startObservingResize()
    }
    if (prevShouldObserve && !shouldObserveResize) {
      this.stopObservingResize()
    }
  }

  componentWillUnmount() {
    this.stopObservingResize()
  }

  startObservingResize = () => {
    const root = this.ref.current
    if (root) {
      this.unlistenResize_ = listenResize(root, this.updateIsFullWidth)
    }
  }

  stopObservingResize() {
    if (this.unlistenResize_) {
      this.unlistenResize_()
    }
  }

  updateHelperImageSrc = () => {
    const {videoWidth, videoHeight} = this.state
    if (!videoWidth || !videoHeight) {
      return
    }
    const [width, height] = reduce(videoWidth, videoHeight)
    const canvas = document.createElement('canvas')
    canvas.width = width
    canvas.height = height
    const helperImageSrc = canvas.toDataURL()
    this.setState({helperImageSrc})
  }

  updateIsFullWidth = () => {
    const {videoWidth, videoHeight} = this.state
    if (!videoWidth || !videoHeight) {
      return
    }
    const root = this.ref.current
    if (!root) return

    const {width, height} = (root as any).getBoundingClientRect()
    // 因为视频缩放后，长宽可能不严格相等，所以认为差值小于等于 0.01 的就算相等。
    // 比如 1280x720 (1.777777778) 和 848x478 (1.774058577)，认为相等。
    const isFullWidth = width / height - videoWidth / videoHeight <= 0.01
    if (isFullWidth !== this.state.isFullWidth) {
      this.setState({isFullWidth})
    }
  }

  triggerUpdateIsFullWidth = () => requestAnimationFrame(this.updateIsFullWidth)

  updateVideoSize = ({videoWidth, videoHeight}: any) => {
    this.setState({videoWidth, videoHeight})
  }

  render() {
    const {children} = this.props
    const {isFullWidth, helperImageSrc} = this.state
    return (
      <PositionContext.Provider
        value={{
          isFullWidth,
          // @ts-expect-error ts-migrate(2322) FIXME: Type '{ isFullWidth: boolean; helperImageSrc: null... Remove this comment to see the full error message
          helperImageSrc,
          updateVideoSize: this.updateVideoSize,
        }}
      >
        {/* @ts-expect-error ts-migrate(2322) FIXME: Type 'RefObject<unknown>' is not assignable to typ... Remove this comment to see the full error message */}
        <div className={css(styles.root)} ref={this.ref}>
          {children}
        </div>
      </PositionContext.Provider>
    )
  }
}
