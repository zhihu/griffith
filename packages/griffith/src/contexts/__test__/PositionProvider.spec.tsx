import React, {useContext, useEffect, useRef} from 'react'
import {act, render} from '@testing-library/react'
import PositionContext, {PositionContextValue} from '../PositionContext'
import PositionProvider from '../PositionProvider'

const waitRAF = () => new Promise(requestAnimationFrame)

test('PositionProvider', async () => {
  let ctx!: PositionContextValue
  const MyComponent = () => {
    const ref = useRef<HTMLDivElement>(null)
    const value = (ctx = useContext(PositionContext))
    useEffect(() => {
      Object.defineProperties(ref.current?.parentNode, {
        getBoundingClientRect: {
          value() {
            return {width: 100, height: 100}
          },
        },
      })
      value.updateVideoSize({videoWidth: 100, videoHeight: 100})
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])
    return <div ref={ref} />
  }

  act(() => {
    render(
      <PositionProvider>
        <MyComponent />
      </PositionProvider>
    )
  })
  expect(ctx).toMatchSnapshot()
  await act(async () => {
    // Provider 内使用了 raf 延迟更新状态
    await waitRAF()
  })
  expect(ctx).toMatchSnapshot()
})
