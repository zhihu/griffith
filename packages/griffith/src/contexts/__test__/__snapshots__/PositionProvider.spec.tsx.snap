// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionProvider 1`] = `
{
  "helperImageSrc": "data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' width='100' height='100'></svg>",
  "isFullWidth": false,
  "updateVideoSize": [Function],
}
`;

exports[`PositionProvider 2`] = `
{
  "helperImageSrc": "data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' width='100' height='100'></svg>",
  "isFullWidth": true,
  "updateVideoSize": [Function],
}
`;
